lib/dialyzer/src/erl_types.erl

```diff
@@ -419,347 +418,86 @@ t_is_none(_) -> false.
-      NewOpaqueTypes = NewOpaque ++ NewOpaqueTypes0,
-      decoration(OpaqueTypes, Type, Opaques, NewOpaqueTypes, NewAll)
-  end;
-decoration([], _Type, _Opaques, NewOpaqueTypes, All) ->
-  {NewOpaqueTypes, All}.
-
--spec list_decorate([erl_type()], [erl_type()], opaques()) -> [erl_type()].
-
-list_decorate(List, L, Opaques) ->
-  [decorate(Elem, E, Opaques) || {Elem, E} <- lists:zip(List, L)].
-
-union_decorate(U1, U2, Opaques) ->
-  Union = union_decorate(U1, U2, Opaques, 0, []),
-  ?untagged_union(A,B,F,I,L,N,T,_,Map) = U1,
-  ?untagged_union(_,_,_,_,_,_,_,Opaque,_) = U2,
-  List = [A,B,F,I,L,N,T,Map],
-  DecList = [Dec ||
-              E <- List,
-              not t_is_none(E),
-              not t_is_none(Dec = decorate(E, Opaque, Opaques))],
-  t_sup([Union|DecList]).
-
-union_decorate([?none|Left1], [_|Left2], Opaques, N, Acc) ->
-  union_decorate(Left1, Left2, Opaques, N, [?none|Acc]);
-union_decorate([T1|Left1], [?none|Left2], Opaques, N, Acc) ->
-  union_decorate(Left1, Left2, Opaques, N+1, [T1|Acc]);
-union_decorate([T1|Left1], [T2|Left2], Opaques, N, Acc) ->
-  union_decorate(Left1, Left2, Opaques, N+1, [decorate(T1, T2, Opaques)|Acc]);
-union_decorate([], [], _Opaques, N, Acc) ->
-  if N =:= 0 -> ?none;
-     N =:= 1 ->
-      [Type] = [T || T <- Acc, T =/= ?none],
-      Type;
-     N >= 2  -> ?union(lists:reverse(Acc))
+%% Returns whether the `Given` type implicitly violates the opacity of opaque
+%% nominals of the `Required` type.
+-spec t_opacity_conflict(Given :: erl_type(),
+                         Required :: erl_type(),
+                         Module :: module()) ->
+    none | expected_opaque | expected_transparent.
+t_opacity_conflict(Given, Required, Module) ->
+  %% Opacity violations are detected by selectively blinding the infimum
+  %% routine to the structure of opaque types that we are not supposed to know
+  %% anything about.
+  %%
+  %% If the infimum of the `Given` and `Required` types is possible, we replace
+  %% the structural component of opaques with a magic value whose infimum with
+  %% anything else becomes `none()`, forcing a failure when the original
+  %% opaques introduce more information.
+  %%
+  %% Conversely, if the infimum of the `Given` and `Required` types is
+  %% impossible, we replace the structural component of opaques with `any()` to
+  %% force success when the altered opaques introduce more information (note
+  %% the inversion).
+  %%
+  %% From there, we can detect opacity violations by checking whether the
+  %% infimum of (blinded `Given`) and (blinded `Required`) is equal to the
+  %% blinded infimum of `Given` and `Required`.
+  Direction = case t_is_impossible(t_inf(Given, Required)) of
+                true -> ?any;
+                false -> ?opaque
+              end,
+
+  RequiredBlind = oc_mark(Required, Direction, Module),
+  GivenBlind = oc_mark(Given, Direction, Module),
+
+  %% If the `Required` type does not change when blinded, we know that the call
```

Comment:
How do we expect this logic to handle complex types which Dialyzer "simplifies", in terms of genuine issues not being raised due to type simplification vs. issues falsely raised due type simplification?

Reply:
We'll only see such simplifications where we have unions of opaques and non-opaques, which is iffy enough that we're adding a  warning for that in this PR, so I don't expect this to become a huge problem.

Reply:
An opaque with inner type `any` is still an opaque. `erl_types:t_opacity_conflict/3` will detect the opaque violation regardless of whether the opaque was simplified by `t_limit/2`. During testing, we had false positives due to our incorrect treatment for nominals in `t_limit/2`. We'll see if new ones pop up.

Reply:
Ah, yes, that's exactly what I was getting at! Sounds like you've considered those cases I was wondering about :)